### PR TITLE
7817 fix api response status 500 instead of 400 when coordinates not specified

### PR DESCRIPTION
**Pull request by Cr1stal423**
GreenCity PR
Issue Link:
[7817](https://github.com/orgs/ita-social-projects/projects/44/views/1?pane=issue&itemId=89212974&issue=ita-social-projects%7CGreenCity%7C7817)
Changes:

    Added logic to validate that coordinates are not empty while creating an event.
    Added new error messages.
    Updated tests related to the new logic.

Current Behavior:

The response now indicates when coordinates are not specified.

Screenshot from 2024-12-02 20-37-45

### CODE REVIEW

File: service/src/main/java/greencity/service/EventServiceImpl.java [hunk 1]
```diff
@@ -601,6 +601,33 @@ private void addNewImages(Event toUpdate, UpdateEventDto updateEventDto, Multipa
...
+            throw new BadRequestException(ErrorMessage.INVALID_COORDINATES);
+        }
+        addAddressToLocation(addEventDtoRequest.getDatesLocations());
+    }
+
+    public boolean validateCoordinates(List<EventDateLocationDto> eventDateLocationDtos) {
+        for (EventDateLocationDto eventDateLocationDto : eventDateLocationDtos) {
+            AddressDto coordinates = eventDateLocationDto.getCoordinates();
+
+            if (coordinates == null || coordinates.getLatitude() == null || coordinates.getLongitude() == null) {
```
Comment (KizerovDmitriy) [inline service/src/main/java/greencity/service/EventServiceImpl.java]: maybe make sens use Objects.isNull() ckeck

File: service/src/main/java/greencity/service/EventServiceImpl.java [hunk 2]
```diff
@@ -601,6 +601,33 @@ private void addNewImages(Event toUpdate, UpdateEventDto updateEventDto, Multipa
...
+                return false;
+            }
+
+            double latitude = coordinates.getLatitude();
+            double longitude = coordinates.getLongitude();
+
+            if (latitude < -90.0 || latitude > 90.0 || longitude < -180.0 || longitude > 180.0) {
```
Comment (KizerovDmitriy) [inline service/src/main/java/greencity/service/EventServiceImpl.java]: maybe move this check to separate method?

File: service/src/test/java/greencity/service/EventServiceImplTest.java [hunk 3]
```diff
@@ -157,6 +157,11 @@ void save() {
         }.getType())).thenReturn(tags);
         when(googleApiService.getResultFromGeoCodeByCoordinates(any()))
             .thenReturn(ModelUtils.getAddressLatLngResponse());
+        AddressDto build = AddressDto.builder()
+            .latitude(ModelUtils.getAddressDto().getLatitude())
+            .longitude(ModelUtils.getAddressDto().getLongitude())
+            .build();
+        when(modelMapper.map(ModelUtils.getAddressLatLngResponse(), AddressDto.class)).thenReturn(build);
```
Comment (KizerovDmitriy) [inline service/src/test/java/greencity/service/EventServiceImplTest.java:161]: move this AddressDto builder to ModelsUtils class

File: service/src/main/java/greencity/service/EventServiceImpl.java (hunk 1 above)
Comment (Cr1stal423) [reply service/src/main/java/greencity/service/EventServiceImpl.java]: Corrected

File: service/src/main/java/greencity/service/EventServiceImpl.java (hunk 2 above)
Comment (Cr1stal423) [reply service/src/main/java/greencity/service/EventServiceImpl.java]: Corrected

File: service/src/test/java/greencity/service/EventServiceImplTest.java (hunk 3 above)
Comment (Cr1stal423) [reply service/src/test/java/greencity/service/EventServiceImplTest.java:161]: Corrected

**Review by nazariusPr — APPROVED** [review]

File: service/src/main/java/greencity/service/EventServiceImpl.java [hunk 4]
```diff
@@ -601,6 +601,38 @@ private void addNewImages(Event toUpdate, UpdateEventDto updateEventDto, Multipa
...
+        checkingEqualityDateTimeInEventDateLocationDto(addEventDtoRequest.getDatesLocations());
+
+        if (!validateCoordinates(addEventDtoRequest.getDatesLocations())) {
+            throw new BadRequestException(ErrorMessage.INVALID_COORDINATES);
+        }
+        addAddressToLocation(addEventDtoRequest.getDatesLocations());
+    }
+
+    private boolean isValidCoordinate(double latitude, double longitude) {
+        return latitude >= -90.0 && latitude <= 90.0 && longitude >= -180.0 && longitude <= 180.0;
```
Comment (holotsvan) [inline service/src/main/java/greencity/service/EventServiceImpl.java]:
```java
return Math.abs(latitude) <= 90.0 && Math.abs(longitude) <= 180.0;
```
what about such solution?

**Review by holotsvan — APPROVED** [review]

File: service/src/main/java/greencity/service/EventServiceImpl.java (hunk 4 above)
Comment (Cr1stal423) [reply service/src/main/java/greencity/service/EventServiceImpl.java]: Changed

File: service-api/src/main/java/greencity/constant/ErrorMessage.java [hunk 5]
```diff
@@ -179,6 +179,9 @@ public class ErrorMessage {
     public static final String USER_HAS_NO_FRIEND_WITH_ID = "User has no friend with this id: ";
     public static final String INVALID_DURATION = "The duration for such habit is lower than previously set";
     public static final String ADDRESS_NOT_FOUND_EXCEPTION = "No address found for the given coordinates.";
+    public static final String INVALID_COORDINATES = "The coordinates field must not be empty";
+    public static final String INVALID_LONGITUDE = "Longitude must be between -180 and 180 degrees";
+    public static final String INVALID_LATITUDE = "Latitude must be between -90 and 90 degrees";
```
Comment (ChernenkoVitaliy) [inline service-api/src/main/java/greencity/constant/ErrorMessage.java:184]: Why do you need constants INVALID_LONGITUDE and INVALID_LATITUDE?
You don't use them anywhere.

File: service/src/main/java/greencity/service/EventServiceImpl.java [hunk 6]
```diff
@@ -601,6 +601,38 @@ private void addNewImages(Event toUpdate, UpdateEventDto updateEventDto, Multipa
         }
     }
 
+    private void validateEventRequest(AddEventDtoRequest addEventDtoRequest) {
+        checkingEqualityDateTimeInEventDateLocationDto(addEventDtoRequest.getDatesLocations());
```
Comment (ChernenkoVitaliy) [inline service/src/main/java/greencity/service/EventServiceImpl.java:604]: You already call this method on line 135 in method save.
I think it is not really necessary to check twice.

**Review by ChernenkoVitaliy — APPROVED** [review]

File: service/src/main/java/greencity/service/EventServiceImpl.java (hunk 6 above)
Comment (Cr1stal423) [reply service/src/main/java/greencity/service/EventServiceImpl.java:604]: Moved method call only to validateEventRequest

**Pull request merged**